webapp/public/serviceworker.js

```diff
@@ -0,0 +1,149 @@
+var CACHE_NAME = 'pxt-cache-v2';
+
+
+
+// keep in sync with release.manifest
+var urlsToCache = [
+  // docs
+  '/doccdn/jquery.js',
+  '/doccdn/semantic.css',
+  '/doccdn/blockly.css',
+  '/docfiles/docs.js',
+  '/docfiles/target.js',
+  '/docfiles/style.css',
+  '/docfiles/target.css',
+  '/docfiles/vs.css',
+  '/--embed',
+  "https://az742082.vo.msecnd.net/pub/psopafpj",
+
+  // editor
+  "/blb/semantic.js",
+  "/blb/main.js",
+  "/blb/typescript.js",
+  "/blb/marked/marked.min.js",
+  "/blb/highlight.js/highlight.pack.js",
+  "/blb/lzma/lzma_worker-min.js",
+  "/blb/fuse.min.js",
+  "/blb/jquery.js",
+  "/blb/pxtlib.js",
+  "/blb/pxtcompiler.js",
+  "/blb/pxtblocks.js",
+  "/blb/pxtwinrt.js",
+  "/blb/pxteditor.js",
+  "/blb/pxtsim.js",
+  "/blb/blockly.css",
+  "/blb/semantic.css",
+  "/blb/rtlsemantic.css",
+  "/blb/custom.css",
+  "/blb/icons.css",
+  "@defaultLocaleStrings@",
+  "@cachedHexFilesArray@",
+  "@targetEditorJs@",
+
+  // Blockly
+  "/blb/blockly/blockly_compressed.js",
+  "/blb/blockly/blocks_compressed.js",
+  "/blb/blockly/msg/js/en.js",
+  "/cdn/blockly/media/sprites.png",
+  "/cdn/blockly/media/click.mp3",
+  "/cdn/blockly/media/disconnect.wav",
+  "/cdn/blockly/media/delete.mp3",
+
+  // monaco
+  "/blb/vs/loader.js",
+  "/blb/vs/base/worker/workerMain.js",
+  "/blb/vs/basic-languages/src/bat.js",
+  "/blb/vs/basic-languages/src/cpp.js",
+  "/blb/vs/editor/editor.main.css",
+  "/blb/vs/editor/editor.main.js",
+  "/blb/vs/editor/editor.main.nls.js",
+  "/blb/vs/language/json/jsonMode.js",
+  "/blb/vs/language/json/jsonWorker.js",
+  "/blb/vs/language/typescript/lib/typescriptServices.js",
+  "/blb/vs/language/typescript/src/mode.js",
+  "/blb/vs/language/typescript/src/worker.js",
+
+  // AI
+  "https://az416426.vo.msecnd.net/scripts/a/ai.0.js"
+];
+
+self.addEventListener('install', function (event) {
+  // Perform install step:  loading each required file into cache
+  event.waitUntil(
+    caches.open(CACHE_NAME).then(function (cache) {
+      // Add all offline dependencies to the cache
+      return cache.addAll(urlsToCache);
+    })
+    .then(function() {
+      // At this point everything has been cached
+      return self.skipWaiting();
+    })
+  );
+})
+self.addEventListener('activate', function(event) {
+  var cacheWhitelist = [CACHE_NAME];
+  event.waitUntil(
+    caches.keys().then(function(cacheNames) {
+      return Promise.all(
+        cacheNames.map(function(cacheName) {
+          if (cacheWhitelist.indexOf(cacheName) === -1) {
+            return caches.delete(cacheName);
+          }
+        })
+      );
+    }).then(function () {
+      // Calling claim() to force a "controllerchange" event on navigator.serviceWorker
+      event.waitUntil(self.clients.claim());
+    })
+  );
+});
+
+self.addEventListener('fetch', function(event) {
+  event.respondWith(
+    (event.request.url.indexOf('/api/') >= 0) ?
+      fetch(event.request)
+    : caches.match(event.request).then(function (response) {
```

Comment:
How are the cache entries invalidated?

Reply:
I think this current cache strategy is no good. 
This is my proposal: 
- static files (index.html, JS and CSS files) are cached on install (like above). 
- doc files in pxtarget menu options (projects, examples), are also cached on install
- docs have a button that is "Available offline" and adds the doc page to the cache
- API calls are never cached and are network only
- App insights is always network only
- General strategy is: for JS / CSS files, query cache first then network 
- For all other files, do a cache and network race.

@mmoskal @pelikhan thoughts?

Reply:
Everything that sits on the CDN is uniquely determined by name. The entries have the form `https://pxt.azureedge.net/blob/<SHA>/<PATH>` or `https://pxt.azureedge.net/commit/<SHA>/<PATH>`. I think, we could just use `<PATH>` to see which entries to invalidate - when we get a request for `https://pxt.azureedge.net/blob/abc...def/foo/bar` and we already have `https://pxt.azureedge.net/blob/123...456/foo/bar` then we remove the existing one and add the new one.